### PR TITLE
Kinesis refactor server abstraction 5265

### PR DESCRIPTION
**Pull request by SwatsonCodes**
This PR refactors the `kinesis mock` and `kinesalite` starters to make use of the `Server` class abstraction as per https://github.com/localstack/localstack/issues/5265.

### CODE REVIEW

File: localstack/services/kinesis/kinesis_starter.py [hunk 1]
```diff
@@ -1,214 +1,66 @@
...
-            def _return_listener(*_):
-                try:
-                    ret_code = PROCESS_THREAD.result_future.result()
-                    if ret_code not in [0, None]:
-                        LOG.error("kinesis terminated with return code %s", ret_code)
-                finally:
-                    kinesis_stopped.set()
-
-            start_thread(_return_listener)
+def start_kinesis(port=None, update_listener=None, asynchronous=None):
```
Comment (SwatsonCodes) [inline localstack/services/kinesis/kinesis_starter.py:14]: this `asynchronous` arg is necessary because it is passed by by the starter [here](https://github.com/localstack/localstack/blob/master/localstack/services/plugins.py#L170). However, this arg is not actually used by this function and I'm not totally sure what it's intent is :)

File: localstack/services/kinesis/kinesis_starter.py [hunk 2]
```diff
@@ -1,214 +1,80 @@
...
-                    ret_code = PROCESS_THREAD.result_future.result()
-                    if ret_code not in [0, None]:
-                        LOG.error("kinesis terminated with return code %s", ret_code)
-                finally:
-                    kinesis_stopped.set()
-
-            start_thread(_return_listener)
+_server: Optional[Server] = None  # server singleton
+
+
```
Comment (SwatsonCodes) [inline localstack/services/kinesis/kinesis_starter.py:13]: I relied on `dynamdb_starter.py` heavily for refactoring this file. Parts of the code such as `_start_kinesis_helper()` and `restart_kinesis()` are nearly copy/pasted from there. To avoid further repetition, it might be worth considering either adding these as abstractions to the base `Server` class, or creating a new abstraction layer over `Server` to hold this logic (e.g. a `ServerStarter` class)

File: localstack/services/kinesis/kinesis_mock_server.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                    latency_param,
+                    kinesis_data_dir_param,
+                    log_level_param,
+                    init_streams_param,
+                    self._bin_path,
+                )
+            )
+        else:
+            chmod_r(self._bin_path, 0o777)
+            cmd = "KINESIS_MOCK_PLAIN_PORT=%s SHARD_LIMIT=%s %s %s %s %s %s --gc=G1" % (
```
Comment (SwatsonCodes) [inline localstack/services/kinesis/kinesis_mock_server.py:88]: There's lots of string interpolation going on here and in the `kinesalite` server as well. I personally find the `%s` syntax hard to follow and would like to create a non-urgent issue for refactoring these instances to use `.format()` instead. Thoughts?

File: localstack/services/install.py [hunk 4]
```diff
@@ -303,31 +304,46 @@ def install_elasticmq():
...
 
-def install_kinesis_mock():
-    target_dir = INSTALL_PATH_KINESIS_MOCK
-
+def get_is_kinesis_mock_installed() -> Tuple[bool, str]:
+    """
+    Checks the host system to see if kinesis mock is installed and where.
+    Returns:
+        True if kinesis mock is installed (False otherwise) and the expected installation path
+    """
```
Comment (SwatsonCodes) [inline localstack/services/install.py:339]: I'm not sure what the overall feeling at LocalStack is around docstrings. I don't see many of them in the codebase, but I find them helpful (and was missing them during while working on this refactor) so I took the time to include some here.
If these are unwanted or should be formatted differently LMK and I'll fix it :)

File: localstack/services/kinesis/kinesis_mock_server.py (hunk 3 above)
Comment (whummer) [reply localstack/services/kinesis/kinesis_mock_server.py:88]: Great point. The `%s` interpolation syntax is mostly for historical reasons (in the past we used to be compatible with python 2 & 3) - these days we tend to use f-strings (or `.format()`) instead. 👍

File: localstack/services/kinesis/kinesis_starter.py [hunk 5]
```diff
@@ -1,214 +1,80 @@
...
     if expect_shutdown:
-        assert out is None or kinesis_stopped.is_set()
+        assert out is None
     else:
-        assert not kinesis_stopped.is_set()
-        assert out and isinstance(out.get("StreamNames"), list)
-
-
-def kinesis_running():
-    return PROCESS_THREAD is not None
```
Comment (alexrashed) [inline localstack/services/kinesis/kinesis_starter.py:203]: Unfortunately, this method is used in one of our extensions, so this method should stay.
Should be fairly easy though, since the `_server` is only set if the server is running.

File: localstack/services/kinesis/kinesis_starter.py [hunk 6]
```diff
@@ -1,214 +1,80 @@
...
-        port=port,
-        backend_port=backend_port,
+def _start_kinesis_helper(server: Server, port=None, update_listener=None) -> Server:
+    server.start()
+    log_startup_message("Kinesis")
+    port = port or config.PORT_KINESIS
+    start_proxy_for_service(
+        "kinesis",
+        port,
+        backend_port=_server.port,
```
Comment (alexrashed) [inline localstack/services/kinesis/kinesis_starter.py:48]: Nitpicking: Since the function is only used in one other function and only consists of a few lines, I think we could just include that in the `start_kinesis` function directly.

**Review by alexrashed — CHANGES_REQUESTED** [review]
Cool changes to further clean up the codebase! They look really clean and nice. The only blocker is the `kinesis_running` function.
With the other comments I'm just nitpicking. ;)
Concerning the docstrings, I think we should stick to restructured text. We are currently thinking about introducing certain code style rules or recommendations (but we don't have anything in place right now).

File: localstack/services/kinesis/kinesis_starter.py (hunk 5 above)
Comment (SwatsonCodes) [reply localstack/services/kinesis/kinesis_starter.py:203]: Thanks for teaching me that the name of that format is `reST`! We used that at my previous employer and I like it better than what I put up here, which is just based on the first example that I pulled up :)

File: localstack/services/install.py (hunk 4 above)
Comment (SwatsonCodes) [reply localstack/services/install.py:339]: Thanks for teaching me that the name of that format is reST! We used that at my previous employer and I like it better than what I put up here, which is just based on the first example that I pulled up :)

**Comment by SwatsonCodes** [comment]
Thanks for the helpful feedback, Nitpicks are always appreciated :)

**Review by calvernaz — APPROVED** [review]
LGTM, the only nit I have is the use of `%` format of strings, I guess the team as a preference for the f'format

**Comment by SwatsonCodes** [comment]
> LGTM, the only nit I have is the use of `%` format of strings, I guess the team as a preference for the f'format

yeah I don't care much for the `%` formatting either. I hope to address it further on down the line.

File: localstack/services/kinesis/kinesis_starter.py [hunk 7]
```diff
@@ -1,214 +1,87 @@
...
-        # giving the process some time to startup; TODO: to be replaced with service lifecycle plugin
-        time.sleep(1)
+    global _server
+    if _server:
+        _server.shutdown()
+        _server.join(timeout=10)
+        _server = None
+
+    LOG.debug("Restarting Kinesis process ...")
+    start_kinesis(update_listener=kinesis_listener)
```
Comment (whummer) [inline localstack/services/kinesis/kinesis_starter.py:87]: @SwatsonCodes Looks like this change was breaking persistence tests in -ext - should probably be this instead:
```
     start_kinesis(update_listener=kinesis_listener.UPDATE_KINESIS)
```
?
